### PR TITLE
i18n: Reference ICU ids from container messages

### LINKED ISSUE #17506: Translate `select` and `plural` loses references in xliff, xliff2 and xmb files
## I'm submitting a ...

<pre><code>
[ ] Regression (behavior that used to work and stopped working in a new release)
[X] Bug report 
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

Translate `select` and `plural` looses references in xliff, xliff2 and xmb files.

I have the following entries in my .html file:

	<span i18n="@@span1">Here we have:
	    {count, plural, =0 {zero} =1 {one}} or {gender, select, m {male} f {female}}
	</span>

	<span i18n="@@span2">Here we have:
	    {gender, select, m {male} f {female}}
	</span>

Note that `span1` and `span2` both use the same `{gender, select, m {male} f {female}}`. The result in xliff now is

      <trans-unit id="span1" datatype="html">
        <source>Here we have:
	         <x id="ICU"/> or <x id="ICU_1"/>
        </source>
        <target/>
        <context-group purpose="location">
          <context context-type="sourcefile">app/dashboard-view/dashboard-view.component.ts</context>
          <context context-type="linenumber">28</context>
        </context-group>
      </trans-unit>
      <trans-unit id="847b0b589338e2d1ec3187b39d01510fce4fe9da" datatype="html">
        <source>{VAR_PLURAL, plural, =0 {zero} =1 {one} }</source>
        <target/>
        <context-group purpose="location">
          <context context-type="sourcefile">app/dashboard-view/dashboard-view.component.ts</context>
          <context context-type="linenumber">29</context>
        </context-group>
      </trans-unit>
      <trans-unit id="14c7055d67771a3b7b6888d282ac092896be06b6" datatype="html">
        <source>{VAR_SELECT, select, m {male} f {female} }</source>
        <target/>
        <context-group purpose="location">
          <context context-type="sourcefile">app/dashboard-view/dashboard-view.component.ts</context>
          <context context-type="linenumber">29</context>
        </context-group>
        <context-group purpose="location">
          <context context-type="sourcefile">app/dashboard-view/dashboard-view.component.ts</context>
          <context context-type="linenumber">34</context>
        </context-group>
      </trans-unit>
      <trans-unit id="span2" datatype="html">
        <source>Here we have:
	         <x id="ICU"/>
        </source>
        <target/>
        <context-group purpose="location">
          <context context-type="sourcefile">app/dashboard-view/dashboard-view.component.ts</context>
          <context context-type="linenumber">33</context>
        </context-group>
      </trans-unit>

The problem is that the reference `<x id="ICU"/>` in `span2` in the second block can't be found. It's not under this block.

From https://angular.io/guide/i18n: 
> The second translation unit, immediately below the first one, contains the select message. Translate that.

In this case:  It's above because it was joined with the translation from `span1` which you can see from the 2 context references in unit `14c7055d67771a3b7b6888d282ac092896be06b6`

## Expected behavior

Add a reference so that the matching translation unit can be found.

## Minimal reproduction of the problem with instructions

See above.

## Please tell us about your environment

<pre><code>
Angular version:  4.2.1
For Tooling issues:
- Node version: 7
- Platform: Mac

Others:
</code></pre>

### PR DESCRIPTION
**Pull request by petebacondarwin**
When extracting i18n messages from templates, ICU messages are split out from the
message that contains them. This can make it difficult in the translation files to match up
the two messages, especially if the ICU is reused in multiple placeholders.

Fixes #17506

Local testing in the AIO project allows us to generate XLIFF that looks like:

```xliff
<?xml version="1.0" encoding="UTF-8" ?>
<xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
  <file source-language="en-US" datatype="plaintext" original="ng2.template">
    <body>
      <trans-unit id="5982288889525183494" datatype="html">
        <source>Before <x id="ICU" equiv-text="{showTopMenu, select, true {blue} false {green} other {red}}" xid="1000974510044941815"/> After</source>
        <context-group purpose="location">
          <context context-type="sourcefile">src/app/app.component.html</context>
          <context context-type="linenumber">7,9</context>
        </context-group>
      </trans-unit>
      <trans-unit id="1000974510044941815" datatype="html">
        <source>{VAR_SELECT, select, true {blue} false {green} other {red}}</source>
        <context-group purpose="location">
          <context context-type="sourcefile">src/app/app.component.html</context>
          <context context-type="linenumber">7</context>
        </context-group>
      </trans-unit>
    </body>
  </file>
</xliff>
```

Note that the `<x id="ICU"...>` has an `xid` attribute that matches the `id` of the ICU message later in the file.

### CODE REVIEW

**Comment by AndrewKushnir** [comment]
@petebacondarwin I've started looking at the change, but I need another day or two to complete the review process. Meanwhile, I've started a [presubmit](http://test/OCL:400862663:BASE:400862680:1633396948102:f90dfcc4) to see if the changes breaks anything.

File: packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_i18n/icu_logic/TEST_CASES.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
     {
       "description": "should handle single icus",
       "inputFiles": [
-        "single_icu"
+        "single_icu.ts"
```
Comment (petebacondarwin) [inline packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_i18n/icu_logic/TEST_CASES.json:7]: This was actually causing this test-case not to be executed!

**Review by AndrewKushnir — COMMENTED** [review]
@petebacondarwin looks great 👍

**Review by AndrewKushnir — APPROVED** [review]
👍

**Review by AndrewKushnir — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
@petebacondarwin presubmits for the changes in this PR are successful + extra testing went well, so this PR is good to go from g3 perspective.

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit c85bcb0c636d65706541d44be62694720f12ad9e.

**Pull request closed without merging**